File: day07/java/michael-simons/Solution.java
package aoc2019;

import static java.util.stream.Collectors.*;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Collection;
import java.util.Deque;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class Solution {

	static class Computer {
		private final Deque<Integer> io = new ArrayDeque<>();

		private final Integer[] memory;

		private Optional<Integer> next;

		static Computer loadProgram(List<Integer> instructions) {
			return new Computer(instructions.toArray(Integer[]::new));
		}

		private Computer(Integer[] memory) {
			this.memory = memory;
			this.next = Optional.of(0);
		}

		private final Map<Integer, BiFunction<Integer, Integer, Integer>> numericOperations = Map.of(
			1, (a, b) -> a + b,
			2, (a, b) -> a * b
		);

		private final Map<Integer, Function<Integer, Boolean>> conditions = Map.of(
			5, (a) -> a != 0,
			6, (a) -> a == 0
		);

		private final Map<Integer, BiFunction<Integer, Integer, Boolean>> comparisons = Map.of(
			7, (a, b) -> a < b,
			8, (a, b) -> a == 0
		);

		public Optional<Integer> run() {

			next = next.flatMap(this::executeInstruction);
			return read();
		}

		public Optional<Integer> read() {
			return Optional.ofNullable(io.poll());
		}

		public void write(Integer in) {
			io.push(in);
		}

		public boolean expectsInput() {
			return next.isPresent();
		}

		private Optional<Integer> executeInstruction(int ptr) {
			int opcode = memory[ptr] % 100;
			int[] modes = extractModes(memory[ptr]);
			return switch (opcode) {
				case 1, 2 -> {
					var p1 = load(memory[++ptr], modes[0]);
					var p2 = load(memory[++ptr], modes[1]);
					store(memory[++ptr], numericOperations.get(opcode).apply(p1, p2));
					yield executeInstruction(++ptr);
				}
				case 3 -> {
					store(memory[++ptr], read().get());
					yield executeInstruction(++ptr);
				}
				case 4 -> {
					write(load(memory[++ptr], modes[0]));
					yield Optional.of(++ptr);
				}
				case 5, 6 -> {
					var p1 = load(memory[++ptr], modes[0]);
					var p2 = load(memory[++ptr], modes[1]);
					yield executeInstruction(conditions.get(opcode).apply(p1) ? p2 : ++ptr);
				}
				case 7, 8 -> {
					var p1 = load(memory[++ptr], modes[0]);
					var p2 = load(memory[++ptr], modes[1]);
					store(memory[++ptr], comparisons.get(opcode).apply(p1, p2) ? 1 : 0);
					yield executeInstruction(++ptr);
				}
				case 99 -> Optional.empty();
				default -> throw new IllegalArgumentException(String.format("Invalid opcode %d", opcode));
			};
		}

		private static int[] extractModes(Integer of) {
			return IntStream.rangeClosed(3, 5)
				.map(i -> of % (int) Math.pow(10, i) / (int) Math.pow(10, i - 1))
				.toArray();
		}

		private Integer load(Integer p, int mode) {
			return mode == 0 ? memory[p] : p;
		}

		private void store(Integer p, Integer v) {
			memory[p] = v;
		}
	}

	private static <T> void swap(T[] v, int i, int j) {
		T t = v[i];
		v[i] = v[j];
		v[j] = t;
	}

	@SuppressWarnings("unchecked")
	private static <T> Stream<List<T>> permutations(Collection<T> v) {
		return permuteImpl((T[]) v.toArray(), v.size());
	}

	/**
	 * See: https://en.wikipedia.org/wiki/Heap%27s_algorithm
	 *
	 * @param v
	 * @param n
	 * @return
	 */
	private static <T> Stream<List<T>> permuteImpl(T[] v, int n) {

		var builder = Stream.<List<T>>builder();
		int c[] = new int[n];

		builder.accept(List.of(v));

		int i = 0;
		while (i < n) {
			if (c[i] < i) {
				if (i % 2 == 0) {
					swap(v, 0, i);
				} else {
					swap(v, c[i], i);
				}
				builder.accept(List.of(v));
				c[i] += 1;
				i = 0;
			} else {
				c[i] = 0;
				i += 1;
			}
		}

		return builder.build();
	}

	public static void main(String... a) throws IOException {

		var instructions = Files.readAllLines(Path.of("input.txt")).stream().flatMap(s -> Arrays.stream(s.split(",")))
			.map(String::trim)
			.map(Integer::parseInt)
			.collect(toList());

		BiFunction<Integer, List<Map.Entry<Computer, Optional<Integer>>>, Integer> runSequence = (initialInput, sequence) -> {
			AtomicInteger input = new AtomicInteger(initialInput);
			for (var computerAndPhase : sequence) {
				var c = computerAndPhase.getKey();
				c.write(input.get());
				computerAndPhase.getValue().ifPresent(c::write);
				input.set(c.run().get());
			}
			return input.get();
		};

		permutations(IntStream.rangeClosed(0, 4).mapToObj(Optional::of).collect(toList()))
			.map(p -> IntStream.rangeClosed(0, p.size() - 1)
				.mapToObj(i -> Map.entry(Computer.loadProgram(instructions), p.get(i))).collect(toList()))
			.parallel()
			.map(s -> runSequence.apply(0, s))
			.max(Integer::compareTo)
			.ifPresent(d -> System.out.println(String.format("Star two %d", d)));

		permutations(IntStream.rangeClosed(5, 9).mapToObj(Optional::of).collect(toList()))
			.parallel()
			.map(phases -> {
				var computers = IntStream.rangeClosed(0, phases.size() - 1)
					.mapToObj(i -> Computer.loadProgram(instructions)).collect(toList());

				var phasesIterator = phases.iterator();
				Supplier<Optional<Integer>> phaseSupplier = () -> phasesIterator.hasNext() ?
					phasesIterator.next() :
					Optional.empty();
				Supplier<Boolean> nextRun = () -> computers.get(computers.size() - 1).expectsInput();

				var lastOutput = new AtomicInteger(0);
				while (nextRun.get()) {
					var sequence = computers.stream().map(c -> Map.entry(c, phaseSupplier.get())).collect(toList());
					lastOutput.set(runSequence.apply(lastOutput.get(), sequence));
				}
				return lastOutput.get();
			})
			.max(Integer::compareTo)
			.ifPresent(d -> System.out.println(String.format("Star two %d", d)));
	}
}

File: day07/scala/michael-simons/Solution.scala
object Solution extends App {

  // tag::newIoClass[]
  class IO(private var input: Seq[Int] = Seq()) {
    def read: Option[Int] = input match {
      case Nil => None
      case l@_ => {
        input = input.tail
        Some(l.head)
      }
    }

    def <<(v: Int) = { // <1>
      input = v +: input
      this
    }
  }

  // end::newIoClass[]

  abstract class Operation {
    def numargs(): Int

    def execute(ptr: Int, args: Seq[(Int, Int)], reader: (((Int, Int)) => Int), writer: (Int, Int) => Unit, stdIo: IO): Int
  }

  case class Exit() extends Operation {
    def numargs(): Int = 0

    def execute(ptr: Int, args: Seq[(Int, Int)], reader: (((Int, Int)) => Int), writer: (Int, Int) => Unit, stdIo: IO): Int = 0
  }

  trait HasOpcode {
    def opcode(): Int
  }

  trait NumericOperation extends HasOpcode {
    val ops = Seq[(Int, Int) => Int](_ + _, _ * _)

    def numargs(): Int = 3

    def opcode(): Int

    def execute(ptr: Int, args: Seq[(Int, Int)], reader: (((Int, Int)) => Int), writer: (Int, Int) => Unit, stdIo: IO): Int = {
      writer(args(2)._1, ops(opcode() - 1)(reader(args(0)), reader(args(1))))
      ptr + numargs() + 1
    }
  }

  case class Add() extends Operation with NumericOperation {

    def opcode() = 1
  }

  case class Mul() extends Operation with NumericOperation {

    def opcode() = 2
  }

  case class ReadInto() extends Operation {

    def numargs(): Int = 1

    def execute(ptr: Int, args: Seq[(Int, Int)], reader: (((Int, Int)) => Int), writer: (Int, Int) => Unit, stdIo: IO): Int = {
      writer(args(0)._1, stdIo.read.get)
      ptr + numargs() + 1
    }
  }

  case class Echo() extends Operation {

    def numargs(): Int = 1

    def execute(ptr: Int, args: Seq[(Int, Int)], reader: (((Int, Int)) => Int), writer: (Int, Int) => Unit, stdIo: IO): Int = {
      stdIo << reader(args(0))
      ptr + numargs() + 1
    }
  }

  trait ConditionalJump extends HasOpcode {
    val cond = Seq[(Int) => Boolean](_ != 0, _ == 0)

    def numargs(): Int = 2

    def opcode(): Int

    def execute(ptr: Int, args: Seq[(Int, Int)], reader: (((Int, Int)) => Int), writer: (Int, Int) => Unit, stdIo: IO): Int = {
      val q = reader(args(0))
      if (cond(opcode() % 5)(q)) {
        reader(args(1))
      } else {
        ptr + numargs() + 1
      }
    }
  }

  case class JumpIfTrue() extends Operation with ConditionalJump {
    def opcode() = 5
  }

  case class JumpIfFalse() extends Operation with ConditionalJump {
    def opcode() = 6
  }

  trait Comparison extends HasOpcode {
    val comp = Seq[(Int, Int) => Boolean](_ < _, _ == _)

    def numargs(): Int = 3

    def execute(ptr: Int, args: Seq[(Int, Int)], reader: (((Int, Int)) => Int), writer: (Int, Int) => Unit, stdIo: IO): Int = {
      val v1 = reader(args(0))
      val v2 = reader(args(1))
      writer(args(2)._1, if (comp(opcode() % 7)(v1, v2)) 1 else 0)
      ptr + numargs() + 1
    }
  }

  case class IsLessThan() extends Operation with Comparison {
    def opcode() = 7
  }

  case class IsEqual() extends Operation with Comparison {
    def opcode() = 8
  }

  object Operation {
    def apply(ptr: Int, memory: Array[Int]) = memory(ptr) % 100 match {
      case 1 => Add()
      case 2 => Mul()
      case 3 => ReadInto()
      case 4 => Echo()
      case 5 => JumpIfTrue()
      case 6 => JumpIfFalse()
      case 7 => IsLessThan()
      case 8 => IsEqual()
      case 99 => Exit()
      case opcode@_ => throw new Exception(s"Invalid opcode $opcode")
    }
  }

  // tag::theComputer[]
  class Computer(private val memory: Array[Int]) { // <1>

    private def extractModes(of: Int) = for (i <- 3 to 5) yield
      of % Math.pow(10, i).toInt / Math.pow(10, i - 1).toInt

    val stdIo = new IO // <2>

    private var next: Option[Int] = Some(0)

    private def load(p: Int, mode: Int) = if (mode == 0) memory(p) else p

    private def store(p: Int, v: Int) = memory(p) = v

    def run(): Option[Int] = {

      val reader = (load _).tupled
      val writer = store _

      def executionInstruction: (Int => Option[Int]) = (ptr: Int) => { // <3>
        val operation = Operation(ptr, memory)
        val modes = extractModes(memory(ptr))
        val args = (1 to operation.numargs).map(i => (memory(ptr + i), modes(i - 1)))
        operation match {
          case Exit() => None
          case op@Echo() => Some(op.execute(ptr, args, reader, writer, stdIo)) // <4>
          case op@_ => {
            val next = op.execute(ptr, args, reader, writer, stdIo)
            executionInstruction(next)
          }
        }
      }

      next = next.flatMap(executionInstruction)
      stdIo.read // <5>
    }

    def expectsInput() = next.isDefined
  }

  object Computer {
    def loadProgram(instructions: Seq[Int]) = new Computer(instructions.toArray) // <6>
  }

  // end::theComputer[]

  val input = io.Source.fromFile("input.txt")
  val instructions = input.getLines().flatMap(_.split(",")).map(_.trim).map(_.toInt).toSeq
  input.close

  // tag::starOne[]
  def runProgram = (input: Int, computerAndPhase: (Computer, Option[Int])) => {
    val c = computerAndPhase._1
    c.stdIo << input
    computerAndPhase._2 match {
      case v@Some(_) => c.stdIo << v.get
      case _ =>
    }
    c.run().get
  } // <1>

  def runWithoutFeedback(instructions: Seq[Int]): Int = {
    val executeSequence = (computers: Seq[(Computer, Option[Int])]) =>
      computers.foldLeft(0)(runProgram) // <2>
    (0 to 4).map(Option(_))
      .permutations // <3>
      .map((1 to 5).map(_ => Computer.loadProgram(instructions)).zip(_)) // <4>
      .map(executeSequence).max // <5>
  }

  val starOne = runWithoutFeedback(instructions)
  println(s"Star one ${starOne}")
  // end::starOne[]

  // tag::starTwo[]
  def runWithFeedback(instructions: Seq[Int]): Int = {
    val executeSequence2 = (phases: Seq[Option[Int]]) => {
      val computers = (1 to 5).map(_ => Computer.loadProgram(instructions))
      val availablePhases = (phases ++: LazyList.continually(None)).iterator // <1>
      val nextRun = () => computers.last.expectsInput()
      var lastOutput = 0
      while (nextRun()) {
        lastOutput = computers
          .zip(availablePhases)
          .foldLeft(lastOutput)(runProgram) // <2>
      }
      lastOutput
    }
    (5 to 9).map(Option(_)).permutations.map(executeSequence2).max
  }

  val starTwo = runWithFeedback(instructions)
  println(s"Star two ${starTwo}")
  // end::starTwo[]
}
